### PR TITLE
feat : upgraded the app with gravity module to SDK v0.50.0

### PR DESCRIPTION
**Pull request by ruthishvitwit**

### CODE REVIEW

File: x/gravity/keeper/attestation.go [hunk 1]
```diff
@@ -27,7 +28,7 @@ func (k Keeper) Attest(
 		return nil, fmt.Errorf("could not find ValAddr for delegate key, should be checked by now")
 	}
 	valAddr := val.GetOperator()
-	if err := sdk.VerifyAddressFormat(valAddr); err != nil {
+	if err := sdk.VerifyAddressFormat([]byte(valAddr)); err != nil {
```
Comment (byte-bandit) [inline x/gravity/keeper/attestation.go:31]: What about using `valAddr.Bytes()`?

Comment (ruthishvitwit) [reply x/gravity/keeper/attestation.go:31]: as valAddr is a string, it's not possible to do it that way

Comment (aleem1314) [reply x/gravity/keeper/attestation.go:31]: This is not the correct way to convert Bech32 string to bytes. Use address codec for converting bech32 to bytes.

**Review by aleem1314 — CHANGES_REQUESTED** [review]
Still there are few things need to be updated.
Check https://github.com/palomachain/paloma/pull/1036#pullrequestreview-1763965986

**Comment by aleem1314** [comment]
@ruthishvitwit resolve conflicts

File: util/liblog/liblog.go [hunk 2]
```diff
@@ -16,6 +16,9 @@ func FromSDKLogger(l log.Logger) Logr {
 func (l *lgwr) Impl() interface{} {
 	return l
 }
+func (l *lgwr) Impl() interface{} {
+	return l
+}
```
Comment (aleem1314) [inline util/liblog/liblog.go:21]:
```suggestion
```

File: app/app.go [hunk 3]
```diff
@@ -181,10 +222,13 @@ var (
...
 		stakingtypes.NotBondedPoolName: {authtypes.Burner, authtypes.Staking},
 		govtypes.ModuleName:            {authtypes.Burner},
+		gravitymoduletypes.ModuleName:  {authtypes.Minter, authtypes.Burner},
 		ibctransfertypes.ModuleName:    {authtypes.Minter, authtypes.Burner},
 		ibcfeetypes.ModuleName:         nil,
 		icatypes.ModuleName:            nil,
-		evmmoduletypes.ModuleName:      {authtypes.Burner, authtypes.Minter},
+		wasmtypes.ModuleName:           {authtypes.Burner},
+		// treasurymoduletypes.ModuleName: {authtypes.Burner, authtypes.Minter},
+		evmmoduletypes.ModuleName: {authtypes.Burner, authtypes.Minter},
```
Comment (aleem1314) [inline app/app.go:231]:
```suggestion
```

File: x/gravity/keeper/attestation.go [hunk 4]
```diff
@@ -36,7 +44,7 @@ func (k Keeper) Attest(
 	// and prevents validators from submitting two claims with the same nonce.
 	// This prevents there being two attestations with the same nonce that get 2/3s of the votes
 	// in the endBlocker.
-	lastEventNonce, err := k.GetLastEventNonceByValidator(ctx, valAddr)
+	lastEventNonce, err := k.GetLastEventNonceByValidator(ctx, sdk.ValAddress(valAddr))
```
Comment (aleem1314) [inline x/gravity/keeper/attestation.go:47]: This is not the correct way to convert bech32 to Hex. Use address codec here too.

File: x/gravity/keeper/attestation.go [hunk 5]
```diff
@@ -69,10 +77,10 @@ func (k Keeper) Attest(
 	if ethClaim.GetEthBlockHeight() == claim.GetEthBlockHeight() {
 
 		// Add the validator's vote to this attestation // TODO : Only do if it's not already there
-		att.Votes = append(att.Votes, valAddr.String())
+		att.Votes = append(att.Votes, valAddr)
 
 		k.SetAttestation(ctx, claim.GetEventNonce(), hash, att)
-		err = k.SetLastEventNonceByValidator(ctx, valAddr, claim.GetEventNonce())
+		err = k.SetLastEventNonceByValidator(ctx, sdk.ValAddress(valAddr), claim.GetEventNonce())
```
Comment (aleem1314) [inline x/gravity/keeper/attestation.go:83]: ditto

File: x/gravity/types/msgs.go [hunk 6]
```diff
@@ -4,9 +4,10 @@ import (
 	"encoding/hex"
 	"fmt"
 
+	sdkerrors "cosmossdk.io/errors"
 	"github.com/cometbft/cometbft/crypto/tmhash"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	errorsmod "github.com/cosmos/cosmos-sdk/types/errors"
```
Comment (aleem1314) [inline x/gravity/types/msgs.go:10]: We should use `cosmossdk.io/errors` for errors

File: util/liblog/liblog.go (hunk 2 above)
Comment (deepan95dev) [reply util/liblog/liblog.go]: changes commited by @ruthishvitwit

File: x/gravity/types/msgs.go (hunk 6 above)
Comment (deepan95dev) [reply x/gravity/types/msgs.go:10]: updated by @ruthishvitwit

File: x/gravity/keeper/attestation.go (hunk 5 above)
Comment (deepan95dev) [reply x/gravity/keeper/attestation.go]: Updated by @ruthishvitwit

File: app/app.go [hunk 7]
```diff
@@ -173,6 +179,41 @@ var (
...
+		// valsetmodule.AppModuleBasic{},
+		wasm.AppModuleBasic{},
+		evm.AppModuleBasic{},
+		gravitymodule.AppModuleBasic{},
+		// palomamodule.AppModuleBasic{},
+		// treasurymodule.AppModuleBasic{},
+		// consensus.AppModuleBasic{},
+		transfer.AppModuleBasic{},
+		ibc.AppModuleBasic{},
+		tendermint.AppModuleBasic{},
```
Comment (aleem1314) [inline app/app.go:212]: Why we need this module?

Comment (deepan95dev) [reply app/app.go]: Replaced `tendermint.AppModuleBasic{}` with `ibtctm.AppModuleBasic{}`  and pushed

File: x/gravity/keeper/test_common.go [hunk 8]
```diff
@@ -298,20 +300,20 @@ func addValidators(t *testing.T, input *TestInput, count int) {
...
 
 		validator, found := input.StakingKeeper.GetValidator(input.Context, addr)
-		require.True(t, found)
+		require.NoError(t, found)
 
 		pubKey, err := validator.ConsPubKey()
 		require.NoError(t, err)
 
-		err = input.ValsetKeeper.AddExternalChainInfo(input.Context, validator.GetOperator(), []*valsettypes.ExternalChainInfo{
+		err = input.ValsetKeeper.AddExternalChainInfo(input.Context, []byte(validator.GetOperator()), []*valsettypes.ExternalChainInfo{
```
Comment (aleem1314) [inline x/gravity/keeper/test_common.go:316]: Use address codec here

File: x/gravity/keeper/msg_server.go [hunk 9]
```diff
@@ -189,11 +193,11 @@ func (k msgServer) confirmHandlerCommon(ctx sdk.Context, ethAddress string, orch
 		return sdkerrors.Wrap(types.ErrInvalid, "validator is unbonded")
 	}
 
-	if err := sdk.VerifyAddressFormat(validator.GetOperator()); err != nil {
+	if err := sdk.VerifyAddressFormat([]byte(validator.GetOperator())); err != nil {
 		return sdkerrors.Wrapf(err, "discovered invalid validator address for orchestrator %v", orchestrator)
 	}
 
-	ethAddressFromStore, found, err := k.GetEthAddressByValidator(ctx, validator.GetOperator(), chainReferenceId)
+	ethAddressFromStore, found, err := k.GetEthAddressByValidator(ctx, []byte(validator.GetOperator()), chainReferenceId)
```
Comment (aleem1314) [inline x/gravity/keeper/msg_server.go:200]: ditto here

File: x/gravity/keeper/msg_server.go [hunk 10]
```diff
@@ -189,11 +193,11 @@ func (k msgServer) confirmHandlerCommon(ctx sdk.Context, ethAddress string, orch
 		return sdkerrors.Wrap(types.ErrInvalid, "validator is unbonded")
 	}
 
-	if err := sdk.VerifyAddressFormat(validator.GetOperator()); err != nil {
+	if err := sdk.VerifyAddressFormat([]byte(validator.GetOperator())); err != nil {
```
Comment (aleem1314) [inline x/gravity/keeper/msg_server.go:196]: ditto here

File: x/gravity/keeper/keeper.go [hunk 11]
```diff
@@ -220,24 +240,46 @@ func (k Keeper) DeserializeValidatorIterator(vals []byte) stakingtypes.ValAddres
...
+func (k Keeper) GetStore(ctx context.Context) storetypes.KVStore {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	return k.storeGetter.Store(sdkCtx)
 }
 
 type GravityStoreGetter struct {
 	storeKey storetypes.StoreKey
 }
 
+func (gsg GravityStoreGetter) GetChainInfo(ctx context.Context, targetChainReferenceID string) (*evmmoduletypes.ChainInfo, error) {
```
Comment (aleem1314) [inline x/gravity/keeper/keeper.go:256]: Why we need these functions?

File: x/gravity/keeper/keeper.go [hunk 12]
```diff
@@ -93,13 +102,19 @@ func NewKeeper(
...
+	if err != nil {
+		return err
+	}
 	feePool.CommunityPool = feePool.CommunityPool.Add(sdk.NewDecCoinsFromCoins(coins...)...)
-	k.DistKeeper.SetFeePool(ctx, feePool)
+	err = k.DistKeeper.FeePool.Set(ctx, feePool)
+	if err != nil {
+		return err
+	}
 	return nil
```
Comment (aleem1314) [inline x/gravity/keeper/keeper.go:118]:
```suggestion
	return k.DistKeeper.FeePool.Set(ctx, feePool)
	
```

File: x/gravity/keeper/keeper.go (hunk 11 above)
Comment (deepan95dev) [reply x/gravity/keeper/keeper.go:256]: Its being used in 3 places
[batch.go](https://github.com/palomachain/paloma/blob/2226e95be60c4972dff44afbbbe7b4cb0ec268c8/x/gravity/keeper/batch.go#L39)
[evidence.go](https://github.com/palomachain/paloma/blob/2226e95be60c4972dff44afbbbe7b4cb0ec268c8/x/gravity/keeper/evidence.go#L39)
[msg_server.go](https://github.com/palomachain/paloma/blob/2226e95be60c4972dff44afbbbe7b4cb0ec268c8/x/gravity/keeper/msg_server.go#L80)

File: x/gravity/keeper/grpc_query.go [hunk 13]
```diff
@@ -171,10 +171,10 @@ func (k Keeper) LastEventNonceByAddr(
 	if !found {
 		return nil, errors.Wrap(types.ErrUnknown, "address")
 	}
-	if err := sdk.VerifyAddressFormat(validator.GetOperator()); err != nil {
+	if err := sdk.VerifyAddressFormat([]byte(validator.GetOperator())); err != nil {
```
Comment (aleem1314) [inline x/gravity/keeper/grpc_query.go:174]: ditto here

File: x/gravity/keeper/grpc_query.go [hunk 14]
```diff
@@ -171,10 +171,10 @@ func (k Keeper) LastEventNonceByAddr(
...
 		return nil, errors.Wrap(err, "invalid validator address")
 	}
-	lastEventNonce, err := k.GetLastEventNonceByValidator(ctx, validator.GetOperator())
+	lastEventNonce, err := k.GetLastEventNonceByValidator(ctx, []byte(validator.GetOperator()))
```
Comment (aleem1314) [inline x/gravity/keeper/grpc_query.go:177]: ditto here

File: x/gravity/keeper/msg_server.go [hunk 15]
```diff
@@ -133,17 +134,20 @@ func (k msgServer) checkOrchestratorValidatorInSet(ctx sdk.Context, orchestrator
 	}
 
 	// return an error if the validator isn't in the active set
-	val := k.StakingKeeper.Validator(ctx, validator.GetOperator())
+	val, err := k.StakingKeeper.Validator(ctx, []byte(validator.GetOperator()))
```
Comment (aleem1314) [inline x/gravity/keeper/msg_server.go:137]: ditto here

File: x/gravity/keeper/keeper.go [hunk 16]
```diff
@@ -32,18 +39,21 @@ type Keeper struct {
 	paramSpace paramtypes.Subspace
 
 	cdc               codec.BinaryCodec // The wire codec for binary encoding/decoding.
+	ProtoUnmarshaler  keeperutil.ProtoUnmarshaler
+	Protomarshaler    keeperutil.ProtoMarshaler
+	Unmarshaler       proto.Unmarshaler
```
Comment (aleem1314) [inline x/gravity/keeper/keeper.go:44]: Why we need these? You can directly use codec. And I don't find where these are initialized.

File: x/gravity/keeper/keeper.go (hunk 11 above)
Comment (deepan95dev) [reply x/gravity/keeper/keeper.go]: removed ununsed functions and pushed

File: x/gravity/keeper/keeper.go (hunk 16 above)
Comment (deepan95dev) [reply x/gravity/keeper/keeper.go]: removed these too

File: app/app.go [hunk 17]
```diff
@@ -120,6 +120,7 @@ import (
 	porttypes "github.com/cosmos/ibc-go/v8/modules/core/05-port/types"
 	ibcexported "github.com/cosmos/ibc-go/v8/modules/core/exported"
 	ibckeeper "github.com/cosmos/ibc-go/v8/modules/core/keeper"
+	ibctm "github.com/cosmos/ibc-go/v8/modules/light-clients/07-tendermint"
```
Comment (aleem1314) [inline app/app.go:123]:
```suggestion
```

File: app/app.go [hunk 18]
```diff
@@ -173,6 +178,41 @@ var (
...
+		// valsetmodule.AppModuleBasic{},
+		wasm.AppModuleBasic{},
+		evm.AppModuleBasic{},
+		gravitymodule.AppModuleBasic{},
+		// palomamodule.AppModuleBasic{},
+		// treasurymodule.AppModuleBasic{},
+		// consensus.AppModuleBasic{},
+		transfer.AppModuleBasic{},
+		ibc.AppModuleBasic{},
+		ibctm.AppModuleBasic{},
```
Comment (aleem1314) [inline app/app.go:211]:
```suggestion
```

File: x/gravity/keeper/keeper_delegate_key.go [hunk 19]
```diff
@@ -13,18 +15,18 @@ import (
...
 	}
 
 	if err := sdk.VerifyAddressFormat(orch); err != nil {
-		ctx.Logger().Error("invalid orch address")
+		liblog.FromSDKLogger(k.Logger(ctx)).Error("invalid orch address")
 		return validator, false, err
 	}
 
-	validator, found = k.StakingKeeper.GetValidator(ctx, valAddr)
+	validator, err = k.StakingKeeper.GetValidator(ctx, valAddr)
```
Comment (aleem1314) [inline x/gravity/keeper/keeper_delegate_key.go:29]: Handle error message

File: x/gravity/keeper/keeper_delegate_key.go [hunk 20]
```diff
@@ -13,18 +15,22 @@ import (
...
-func (k Keeper) GetOrchestratorValidator(ctx sdk.Context, orch sdk.AccAddress) (validator stakingtypes.Validator, found bool, err error) {
+func (k Keeper) GetOrchestratorValidator(ctx context.Context, orch sdk.AccAddress) (validator stakingtypes.Validator, found bool, err error) {
 	valAddr := sdk.ValAddress(orch)
 	if valAddr == nil {
 		return validator, false, fmt.Errorf("nil valAddr")
 	}
 
 	if err := sdk.VerifyAddressFormat(orch); err != nil {
-		ctx.Logger().Error("invalid orch address")
+		liblog.FromSDKLogger(k.Logger(ctx)).Error("invalid orch address")
```
Comment (byte-bandit) [inline x/gravity/keeper/keeper_delegate_key.go:25]: You're not including the actual error in the logs. Use `WithError(err)` here and all other cases like this.

File: x/gravity/keeper/keeper.go [hunk 21]
```diff
@@ -55,7 +57,7 @@ func NewKeeper(
 	stakingKeeper types.StakingKeeper,
 	bankKeeper types.BankKeeper,
 	slashingKeeper types.SlashingKeeper,
-	distributionKeeper types.DistributionKeeper,
+	distributionKeeper distrkeeper.Keeper,
```
Comment (aleem1314) [inline x/gravity/keeper/keeper.go:60]: Why this change? The type should be from gravity module.

Comment (ruthishvitwit) [reply x/gravity/keeper/keeper.go:60]: cz we were unable to import get or set functions when we tried to use them ...so to call feepool.get/set functions we had to make this change

Comment (taariq) [reply x/gravity/keeper/keeper.go:60]: @byte-bandit

Comment (aleem1314) [reply x/gravity/keeper/keeper.go:60]: > cz we were unable to import get or set functions when we tried to use them ...so to call feepool.get/set functions we had to make this change

I see, distribution keeper has SDK collection fields.

**Review by aleem1314 — APPROVED** [review]
lgtm

**Review by aleem1314 — APPROVED** [review]

File: x/gravity/keeper/keeper.go (hunk 21 above)
Comment (byte-bandit) [reply x/gravity/keeper/keeper.go:60]: Yupp looks good.

File: x/gravity/keeper/attestation_handler.go [hunk 22]
```diff
@@ -57,13 +60,7 @@ func (a AttestationHandler) handleSendToPaloma(ctx sdk.Context, claim types.MsgS
...
 		}
 
-		a.keeper.Logger(ctx).Error("Invalid SendToPaloma receiver",
-			"address", receiverAddress,
-			"cause", addressErr.Error(),
-			"claim type", claim.GetType(),
-			"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
-			"nonce", fmt.Sprint(claim.GetEventNonce()),
-		)
+		liblog.FromSDKLogger(a.keeper.Logger(ctx)).WithFields("address", receiverAddress, "cause", addressErr.Error(), "claim type", claim.GetType(), "id", types.GetAttestationKey(claim.GetEventNonce(), hash), "nonce", fmt.Sprint(claim.GetEventNonce())).Error("Invalid SendToPaloma receiver")
```
Comment (byte-bandit) [inline x/gravity/keeper/attestation_handler.go:63]: I don't care if we do this now of when we merge back into `master`, but line length should not exceed past some sensible limit.

**Review by byte-bandit — DISMISSED** [review]
A lot of stuff that needs improving, but was already lacking before and isn't part of this scope. Only one small remark, mostly looks good.

File: x/gravity/keeper/attestation_handler.go (hunk 22 above)
Comment (ruthishvitwit) [reply x/gravity/keeper/attestation_handler.go]: modified and pushed

**Review by byte-bandit — APPROVED** [review]
LGTM, final review to @aleem1314

**Review by aleem1314 — DISMISSED** [review]

**Review by aleem1314 — APPROVED** [review]

**Review by taariq — APPROVED** [review]
Approving after @byte-bandit and @aleem1314
